src/tree/gpu_hist/gradient_based_sampler.cu

```diff
@@ -0,0 +1,225 @@
+  XGBOOST_DEVICE bool operator()(const GradientPair& gpair) const {
+    return gpair.GetGrad() != 0 || gpair.GetHess() != 0;
+  }
+};
+
+/*! \brief A functor that clears the row indexes with empty gradient. */
+struct ClearEmptyRows : public thrust::binary_function<GradientPair, size_t, size_t> {
+  XGBOOST_DEVICE size_t operator()(const GradientPair& gpair, size_t row_index) const {
+    if (gpair.GetGrad() != 0 || gpair.GetHess() != 0) {
+      return row_index;
+    } else {
+      return SIZE_MAX;
+    }
+  }
+};
+
+GradientBasedSample GradientBasedSampler::SequentialPoissonSampling(
+    common::Span<xgboost::GradientPair> gpair, DMatrix* dmat) {
+  // Transform the gradient to weight = random(0, 1) / abs(grad).
+  thrust::transform(dh::tbegin(gpair), dh::tend(gpair),
+                    thrust::counting_iterator<size_t>(0),
+                    dh::tbegin(row_weight_),
+                    CalculateWeight(common::GlobalRandom()()));
+  return WeightedSampling(gpair, dmat);
+}
+
+// Perform sampling after the weights are calculated.
+GradientBasedSample GradientBasedSampler::WeightedSampling(
+      common::Span<xgboost::GradientPair> gpair, DMatrix* dmat) {
+  // Sort the gradient pairs and row indexes by weight.
+  thrust::sort_by_key(dh::tbegin(row_weight_), dh::tend(row_weight_),
+                      thrust::make_zip_iterator(thrust::make_tuple(dh::tbegin(gpair),
+                                                                   dh::tbegin(row_index_))));
+
+  // Clear the gradient pairs not included in the sample.
+  thrust::fill(dh::tbegin(gpair) + sample_rows_, dh::tend(gpair), GradientPair());
+
+  // Mask the sample rows.
+  thrust::fill(dh::tbegin(sample_row_index_), dh::tbegin(sample_row_index_) + sample_rows_, 1);
+  thrust::fill(dh::tbegin(sample_row_index_) + sample_rows_, dh::tend(sample_row_index_), 0);
+
+  // Sort the gradient pairs and sample row indexes by the original row index.
+  thrust::sort_by_key(dh::tbegin(row_index_), dh::tend(row_index_),
+                      thrust::make_zip_iterator(thrust::make_tuple(dh::tbegin(gpair),
+                                                                   dh::tbegin(sample_row_index_))));
+
+  // Compact the non-zero gradient pairs.
+  thrust::copy_if(dh::tbegin(gpair), dh::tend(gpair), dh::tbegin(gpair_), IsNonZero());
+
+  // Index the sample rows.
+  thrust::exclusive_scan(dh::tbegin(sample_row_index_), dh::tend(sample_row_index_),
+                         dh::tbegin(sample_row_index_));
+  thrust::transform(dh::tbegin(gpair), dh::tend(gpair),
+                    dh::tbegin(sample_row_index_),
+                    dh::tbegin(sample_row_index_),
+                    ClearEmptyRows());
+
+  // Compact the ELLPACK pages into the single sample page.
+  thrust::fill(dh::tbegin(page_->gidx_buffer), dh::tend(page_->gidx_buffer), 0);
+  for (auto& batch : dmat->GetBatches<EllpackPage>(batch_param_)) {
+    page_->Compact(batch_param_.gpu_id, batch.Impl(), sample_row_index_);
```

Comment:
is there some benefit to bailing out sooner once we have the target sample number of rows compacted in, by having the compact method return number of rows compacted in this batch, and using that to decide? if these samples are going to be distributed near evenly across all pages, this may not help.

Reply:
It seems unlikely we won't have a row in a 32MB page.